addons/mrp_subcontracting_purchase/models/stock_move.py

```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models
+
+
+class StockMove(models.Model):
+    _inherit = 'stock.move'
+
+    def _is_purchase_return(self):
+        res = super(StockMove, self)._is_purchase_return()
+        return res or self._is_subcontract_return()
+
+    def _is_subcontract_return(self):
+        self.ensure_one()
+        return not self.is_subcontract and self.warehouse_id \
+            and self.location_id.usage != "supplier" \
+            and self.location_dest_id.usage == "internal" \
+            and self.location_dest_id not in self.env["stock.location"].search([("id", "child_of", self.warehouse_id.view_location_id.id)])
```

Comment:
I would still move that part in `mrp_subcontracting` as nothing depends on `purchase` here